### PR TITLE
[chore] use github.sha over github.ref_head for merge queue group concurrency

### LINKED ISSUE #11841: Concurrency control does not work with merge queue
If I try to add two PRs to the merge queue (like I just tried with #11833 and #11827) I get the following error:

>  Canceling since a higher priority waiting request for 'contrib-tests-' exists

(see e.g. https://github.com/open-telemetry/opentelemetry-collector/actions/runs/12259930519).

This happens because `${{ github.head_ref }}` is empty in merge queue events here: https://github.com/open-telemetry/opentelemetry-collector/blob/cc1d61e5e055018471c970809cd0ed23d56374e0/.github/workflows/contrib-tests.yml#L13-L15

(Thanks @jade-guiton-dd for looking into this!)

To solve this we can do one of the following:
- Use ~~`github.ref`~~ `github.sha` (Jade thinks this would work, I would try this one first)
- Use `github.head_ref` or `github.event.merge_group.head_ref` depending on the `github.event_name`

### PR DESCRIPTION
**Pull request by atoulme**
#### Link to tracking issue
Fixes #11841

### CODE REVIEW

File: .github/workflows/build-and-test-arm.yml [hunk 1]
```diff
@@ -18,7 +18,7 @@ permissions:
 
 # Do not cancel this workflow on main. See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16616
 concurrency:
-  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  group: ${{ github.workflow }}-${{ github.sha || github.run_id }}
```
Comment (dmitryax) [inline .github/workflows/build-and-test-arm.yml:21]: `github.sha` is always present, so we don't need `github.run_id` right?

Comment (atoulme) [reply .github/workflows/build-and-test-arm.yml]: sounds good, removed.

File: .github/workflows/build-and-test-arm.yml [hunk 2]
```diff
@@ -18,7 +18,7 @@ permissions:
 
 # Do not cancel this workflow on main. See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16616
 concurrency:
-  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  group: ${{ github.workflow }}-${{ github.sha }}
```
Comment (jade-guiton-dd) [inline .github/workflows/build-and-test-arm.yml:21]:
```suggestion
  group: ${{ github.workflow }}-${{ github.ref_name }}
```

File: .github/workflows/build-and-test-windows.yaml [hunk 3]
```diff
@@ -9,7 +9,7 @@ on:
   pull_request:
 
 concurrency:
-  group: ${{ github.workflow }}-${{ github.head_ref }}
+  group: ${{ github.workflow }}-${{ github.sha }}
```
Comment (jade-guiton-dd) [inline .github/workflows/build-and-test-windows.yaml:12]:
```suggestion
  group: ${{ github.workflow }}-${{ github.ref_name }}
```

File: .github/workflows/build-and-test.yml [hunk 4]
```diff
@@ -12,7 +12,7 @@ permissions:
   contents: read
 
 concurrency:
-  group: ${{ github.workflow }}-${{ github.head_ref }}
+  group: ${{ github.workflow }}-${{ github.sha }}
```
Comment (jade-guiton-dd) [inline .github/workflows/build-and-test.yml:15]:
```suggestion
  group: ${{ github.workflow }}-${{ github.ref_name }}
```

File: .github/workflows/builder-integration-test.yaml [hunk 5]
```diff
@@ -20,7 +20,7 @@ on:
     types: [checks_requested]
 
 concurrency:
-  group: ${{ github.workflow }}-${{ github.head_ref }}
+  group: ${{ github.workflow }}-${{ github.sha }}
```
Comment (jade-guiton-dd) [inline .github/workflows/builder-integration-test.yaml:23]:
```suggestion
  group: ${{ github.workflow }}-${{ github.ref_name }}
```

File: .github/workflows/changelog.yml [hunk 6]
```diff
@@ -13,7 +13,7 @@ on:
     types: [checks_requested]
 
 concurrency:
-  group: ${{ github.workflow }}-${{ github.head_ref }}
+  group: ${{ github.workflow }}-${{ github.sha }}
```
Comment (jade-guiton-dd) [inline .github/workflows/changelog.yml:16]:
```suggestion
  group: ${{ github.workflow }}-${{ github.ref_name }}
```

File: .github/workflows/check-links.yaml [hunk 7]
```diff
@@ -7,7 +7,7 @@ on:
     types: [checks_requested]
 
 concurrency:
-  group: ${{ github.workflow }}-${{ github.head_ref }}
+  group: ${{ github.workflow }}-${{ github.sha }}
```
Comment (jade-guiton-dd) [inline .github/workflows/check-links.yaml:10]:
```suggestion
  group: ${{ github.workflow }}-${{ github.ref_name }}
```

File: .github/workflows/codeql-analysis.yml [hunk 8]
```diff
@@ -5,7 +5,7 @@ on:
   pull_request:
 
 concurrency:
-  group: ${{ github.workflow }}-${{ github.head_ref }}
+  group: ${{ github.workflow }}-${{ github.sha }}
```
Comment (jade-guiton-dd) [inline .github/workflows/codeql-analysis.yml:8]:
```suggestion
  group: ${{ github.workflow }}-${{ github.ref_name }}
```

File: .github/workflows/contrib-tests.yml [hunk 9]
```diff
@@ -11,7 +11,7 @@ on:
     types: [checks_requested]
 
 concurrency:
-  group: ${{ github.workflow }}-${{ github.head_ref }}
+  group: ${{ github.workflow }}-${{ github.sha }}
```
Comment (jade-guiton-dd) [inline .github/workflows/contrib-tests.yml:14]:
```suggestion
  group: ${{ github.workflow }}-${{ github.ref_name }}
```

**Review by jade-guiton-dd — COMMENTED** [review]
If we want to preserve the old behavior where pushing commits on a PR multiple times would result in previous CI runs being cancelled (to avoid wasting time testing obsolete commits), I think we may want to use `github.ref` or `github.ref_name` instead of `github.sha`. (As far as I can tell all three are defined in both merge groups and PRs, although it's not very well documented.)

**Comment by mx-psi** [comment]
My bad, I suggested using `github.sha` on the linked issue. Let's try `github.ref_name` :)

**Comment by mx-psi** [comment]
Kicked out of the merge queue due to #11847, filed #11848 to fix this and added to the merge queue again

**Pull request merged**